src/storage/src/hummock/iterator/skip_watermark.rs

```diff
@@ -0,0 +1,402 @@
+            inner,
+            remain_watermarks: VecDeque::new(),
+            watermarks,
+        }
+    }
+
+    fn reset_watermark(&mut self) {
+        self.remain_watermarks = self
+            .watermarks
+            .iter()
+            .flat_map(|(table_id, read_watermarks)| {
+                read_watermarks
+                    .vnode_watermarks
+                    .iter()
+                    .map(|(vnode, watermarks)| {
+                        (
+                            *table_id,
+                            *vnode,
+                            read_watermarks.direction,
+                            watermarks.clone(),
+                        )
+                    })
+            })
+            .collect();
+    }
+
+    /// Advance watermark until no watermark remains or the first watermark can possibly
+    /// filter out the current or future key.
+    ///
+    /// Return a flag indicating whether the current key will be filtered by the current watermark.
+    fn advance_watermark(&mut self) -> bool {
+        if self.inner.is_valid() {
+            let key = self.inner.key();
+            let key_table_id = key.user_key.table_id;
+            let (key_vnode, inner_key) = key.user_key.table_key.split_vnode();
+            while let Some((table_id, vnode, direction, watermark)) = self.remain_watermarks.front()
+            {
+                match (table_id, vnode).cmp(&(&key_table_id, &key_vnode)) {
+                    Ordering::Less => {
+                        self.remain_watermarks.pop_front();
+                        continue;
+                    }
+                    Ordering::Equal => {
+                        match direction {
+                            WatermarkDirection::Ascending => {
+                                match inner_key.cmp(watermark.as_ref()) {
+                                    Ordering::Less => {
+                                        // The current key will be filtered by the watermark.
+                                        // Return true to further advance the key.
+                                        return true;
+                                    }
+                                    Ordering::Equal | Ordering::Greater => {
+                                        // The current key has passed the watermark.
+                                        // Advance the next watermark.
+                                        self.remain_watermarks.pop_front();
+                                        // Since it is impossible for a (table_id, vnode) tuple to have multiple
+                                        // watermark, after the pop_front, the next (table_id, vnode) must have
+                                        // exceeded the current key, and we can directly return and mark that the
+                                        // current key is not filtered by the watermark at the front.
+                                        #[cfg(debug_assertions)]
```

Comment:
https://github.com/risingwavelabs/risingwave/pull/13429#discussion_r1405836132 is fixed here.